Added KMS Driver developer guide (#732)
digital-asset/docs.daml.com: docs/2.10.0/docs/canton/usermanual/kms/kms_driver_guide.rst

   SHA256 asymmetric encryption.
-  Signing: Sign data with a key from the KMS and the specified
   algorithm.
-  Asymmetric Decryption: Decrypt a ciphertext with a private key from
   the KMS and the specified algorithm.
-  Symmetric Encryption and Decryption: Encrypt or decrypt with a
   symmetric encryption key from the KMS. The default symmetric
   encryption algorithm of the KMS is used.
-  Key Management: Get the public key of a private key stored in the
   KMS, check if a key exists, and delete a key.
-  Health: Return the health of the KMS Driver instance.

The API is designed as an asynchronous API using Futures. An
OpenTelemetry trace context is passed in from Canton into the KMS driver
operations to be able to link a Canton request to operations in the KMS.
The driver is responsible for propagating the trace context into the
KMS.

Concretely the Scala interface is defined as the following:

.. literalinclude:: /canton/includes/mirrored/enterprise/kms-driver-api/src/main/scala/com/digitalasset/canton/crypto/kms/driver/api/v1/KmsDriver.scala
   :language: Scala

Error Handling and Health
~~~~~~~~~~~~~~~~~~~~~~~~~

> In case the driver experiences an error the Future of the operation should be

Reviewer: should fail

Author: actually the future should be completed as a failure with the exception, so I think `should be failed` is correct here.

failed with a ``KmsDriverException``. When the exception's flag retryable is

> true the caller side, i.e., Canton, will perform a retry with exponential

Reviewer: flag retryable is true, (add a comma)

driver-specific means.

Bundle your driver into a self-contained jar, that is, with all required
libraries included in the jar. That way you only need a single driver jar when
starting Canton with your KMS Driver.

KMS Driver Testing
~~~~~~~~~~~~~~~~~~

The reusable test suite for KMS Drivers is published at
`canton-kms-driver-testing
<https://digitalasset.jfrog.io/ui/repos/tree/General/canton-kms-driver-testing>`__.
Configure your build system to depend on this maven artifact in the test scope
of your project (e.g. for sbt append % Test to limit the dependency to the test
scope).

KmsDriverTest
^^^^^^^^^^^^^

The main part of the test suite is the ``KmsDriverTest`` that tests the
functionality of a driver against the ``KmsDriver`` API.

In the simplest form the specific driver test class extends the
``KmsDriverTest`` and allows the generation of new keys as part of the test:

.. literalinclude:: /canton/includes/mirrored/enterprise/aws-kms-driver/src/test/scala/com/digitalasset/canton/nightly/AwsKmsDriverTest.scala
   :language: Scala
   :start-after: user-manual-entry-begin: AwsKmsDriverTest
   :end-before: user-manual-entry-end: AwsKmsDriverTest

> Given that generation of new keys can be expensive when running tests during
> development frequently, in particular when using Cloud-based KMSs, the test
> suite can also be configured to use predefined keys to test most parts (except
> key generation obviously) of the KMS Driver API:

Reviewer 2: ```suggestion
Generating new keys can be expensive when running tests during
development, in particular when using cloud-based KMSs. To mitigate this, the test
suite can also be configured to use predefined keys to test most parts (except
key generation) of the KMS Driver API:
```

.. literalinclude:: /canton/includes/mirrored/enterprise/aws-kms-driver/src/test/scala/com/digitalasset/canton/nightly/AwsKmsDriverTest.scala
   :language: Scala
   :start-after: user-manual-entry-begin: AwsKmsDriverWithPredefinedKeysTest
   :end-before: user-manual-entry-end: AwsKmsDriverWithPredefinedKeysTest

> For each supported signing/encryption key specification an existing key alias/id
> can be configured as part of the predefined keys maps. When running the test
> suite the generation of new keys is not allowed.

Reviewer 2: ```suggestion
For each supported signing/encryption key specification an existing key alias/ID
can be configured as part of the predefined key maps. When running the test
suite the generation of new keys is not allowed.
```

Author: there can be multiple keys in each map, so I kept the plural and only upper-cased the ID.

KmsDriverFactoryTest
^^^^^^^^^^^^^^^^^^^^

> The test suite for the KMS driver factory is structured similarly:

Reviewer 2: ```suggestion
The test suite for the KMS driver factory is structured similarly to the above:
```

.. literalinclude:: /canton/includes/mirrored/enterprise/aws-kms-driver/src/test/scala/com/digitalasset/canton/nightly/AwsKmsDriverTest.scala
   :language: Scala
   :start-after: user-manual-entry-begin: AwsKmsDriverFactoryTest
   :end-before: user-manual-entry-end: AwsKmsDriverFactoryTest

> The ``KmsDriverFactory`` can write the driver-specific configuration with a
> confidential flag being true, which means any sensitive information in the
> configuration such as credentials, should be omitted from the written
> configuration. A specific test case should be added if your driver-specific
> configuration contains any confidential information, which asserts that the
> sensitive information is omitted.

Reviewer 2: ```suggestion
The ``KmsDriverFactory`` can write the driver-specific configuration with a
confidential flag being true, which means any sensitive information in the
configuration such as credentials should be omitted from the written
configuration. A specific test case should be added if your driver-specific
configuration contains any confidential information, asserting that the
sensitive information is omitted.
```

Run Canton with a KMS Driver
----------------------------

> Configure Canton to run with a KMS driver, for example, with for a

Reviewer: `with for` sounds strange, probably just `for`

participant participant1:

.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/aws-kms-driver.conf
   :language: none
   :start-after: user-manual-entry-begin: AwsKmsDriverConfig
   :end-before: user-manual-entry-end: AwsKmsDriverConfig

Run Canton with your driver jar on its class path:

``java -cp driver.jar:canton.jar com.digitalasset.canton.CantonEnterpriseApp -c canton.conf # further canton arguments``

> Where canton.jar depends on the canton version, e.g.,

Reviewer 2: ```suggestion
Where canton.jar depends on the Canton version, e.g.,
```
